### PR TITLE
LibDSP+Piano: Move Piano's DSP fully to LibDSP

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
Road to #6528 step 3. Piano is now free of DSP code except for sequencing, which will be refactored in the next step.

- [x] Master processor (@Granddave )
- [x] "Classic" synth processor that replicates Piano's old synth engine
- [x] Make both processors work with Piano (master postponed)
- [x] UI integration

Auxiliary changes:
* Several volume type conversion functions in LibAudio along with a move of Frame into its own file (@Granddave )
* clear_with_capacity in Hash{Table, Map} (@Hendiadyoin1 )
* ProcessorEnumParameter (LibDSP) for discrete enum-based parameters. Used for the Classic synth waveform.
* ProcessorEnumWidget (Piano) for controlling ProcessorEnumParameter

Known issues:
* Note data doesn't carry over when the loop repeats
* Multiple synths (one for each track) are broken for now

Note that the sampler is removed for the time being, it will be re-added as its own processor once we have the opportunity to (at least programmatically) swap out processors.

Thanks to @Granddave and @Hendiadyoin1 for their work.

### CODE REVIEW

**Comment by kleinesfilmroellchen** [comment]
> I have no idea about anything audio, apart from knowing live-mixing... The code looks good for what i can tell, I have a few grammar nitpics. It builds and runs on my machine, so ship it! ;)
> 
> Few things I noticed while using Piano:
> 
> * It is very quiet, I have to ramp up my volume all the way to hear anything.
> * The sliders for the synth are not very polished. Apart from me not having much of a clue what they are supposed to do, the values seem to be too large to allow for fine control. Isn't this where all the logarithmic shenanigans are supposed to come in? (Later maybe?)
> * If the value for decay is above 10000 there is no text displayed - Maybe make the labels only show fractional parts if the value is < 100 or sth.

1. It was intended to give more headroom to Piano but I guess we can turn it up until the new engine is in.
2. Logarithmic sliders are among the next steps, yes. For what they do, I feel like the labels are descriptive enough for small LibGUI labels :^), but have a look at ADSR envelopes and digital synthesizer basics if you want to know more.
3. Will do immediately.

File: Userland/Libraries/LibAudio/Buffer.cpp [hunk 1]
```diff
@@ -45,7 +45,7 @@ i32 Buffer::allocate_id()
 }
 
 template<typename SampleReader>
-static void read_samples_from_stream(InputMemoryStream& stream, SampleReader read_sample, Vector<Frame>& samples, int num_channels)
+static void read_samples_from_stream(InputMemoryStream& stream, SampleReader read_sample, Vector<Sample>& samples, int num_channels)
 {
     double norm_l = 0;
     double norm_r = 0;
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibAudio/Buffer.cpp:51]: That is old code, but I'm here anyways :^)

File: Userland/Libraries/LibDSP/Synthesizers.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+// We resample the noise twice per cycle.
+double Classic::noise_position(u8 note)
+{
+    double spc = samples_per_cycle(note);
+    return (get_random_from_seed(round(m_transport->time() / (spc / 2)) * (spc / 2)) - .5) * 2;
+}
+
+double Classic::get_random_from_seed(u64 seed)
+{
+    u64 const a = 6364136223846793005;
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibDSP/Synthesizers.cpp:145]: Oh, I just punched my keyboard.

**Comment by awesomekling** [comment]
This looks like a snowball of unrelated changes. Are all of these commits actually required for the stated goal of "Move Piano's DSP fully to LibDSP"?

**Comment by kleinesfilmroellchen** [comment]
Not all, but most of them. The master processor relies on most of the LibAudio changes, but I can take out these commits if necessary. @Granddave would have to add them to another PR then.

**Comment by kleinesfilmroellchen** [comment]
@awesomekling I have now extracted about nine commits that are not necessary right now, but the other ones are actual "volume theory" necessary for the new processors. I hope this is better now. @Granddave please look for the missing commits and add them back into a new PR.

**Comment by kleinesfilmroellchen** [comment]
I understand that it feels like a snowball, but there are a bunch of refactors here that made the later work easier (and will make future work easier). Additionally, there are just some bug fixes like the LibGUI stuff which were exposed by the "primary" work. Yakshaving has to be done somewhere :^)

**Comment by kleinesfilmroellchen** [comment]
I'll pick this PR apart now, therefore closing it.

**Pull request closed without merging**